fdbserver/tester.actor.cpp

```diff
@@ -271,104 +272,189 @@ Standalone<VectorRef<KeyValueRef>> checkAllOptionsConsumed(VectorRef<KeyValueRef
+			        .detail("Name", workloadName)
+			        .detail("Remaining", *wCount)
+			        .detail("Phase", "End");
+			    return Void();
+		    },
+		    workloads[i]->start(cx)));
+	}
+	auto done = waitForAll(all);
+	if (failureInjection.empty()) {
+		return done;
+	}
+	std::vector<Future<Void>> res;
+	res.reserve(failureInjection.size());
+	for (auto& f : failureInjection) {
+		res.push_back(f->startInjectionWorkload(cx, done));
+	}
+	return waitForAll(res);
+}
+Future<bool> CompoundWorkload::check(Database const& cx) {
+	std::vector<Future<bool>> all;
+	all.reserve(workloads.size());
+	auto wCount = std::make_shared<unsigned>(0);
+	for (int i = 0; i < workloads.size(); i++) {
+		++(*wCount);
+		std::string workloadName = workloads[i]->description();
+		TraceEvent("WorkloadCheckStatus")
+		    .detail("Name", workloadName)
+		    .detail("Count", *wCount)
+		    .detail("Phase", "Start");
+		all.push_back(fmap(
+		    [workloadName, wCount](bool ret) {
+			    --(*wCount);
+			    TraceEvent("WorkloadCheckStatus")
+			        .detail("Name", workloadName)
+			        .detail("Remaining", *wCount)
+			        .detail("Phase", "End");
+			    return true;
+		    },
+		    workloads[i]->check(cx)));
+	}
+	auto done = allTrue(all);
+	if (failureInjection.empty()) {
+		return done;
+	}
+	std::vector<Future<bool>> res;
+	res.reserve(failureInjection.size());
+	for (auto& f : failureInjection) {
+		res.push_back(f->checkInjectionWorkload(cx, done));
+	}
+	return allTrue(res);
+}
+
+ACTOR Future<std::vector<PerfMetric>> getMetricsCompoundWorkload(CompoundWorkload* self) {
+	state std::vector<Future<std::vector<PerfMetric>>> results;
+	for (int w = 0; w < self->workloads.size(); w++) {
+		std::vector<PerfMetric> p;
+		results.push_back(self->workloads[w]->getMetrics());
+	}
+	wait(waitForAll(results));
+	std::vector<PerfMetric> res;
+	for (int i = 0; i < results.size(); ++i) {
+		auto const& p = results[i].get();
+		for (auto const& m : p) {
+			res.push_back(m.withPrefix(self->workloads[i]->description() + "."));
 		}
-		return res;
 	}
+	return res;
+}
 
-	Future<std::vector<PerfMetric>> getMetrics() override { return getMetrics(this); }
-	double getCheckTimeout() const override {
-		double m = 0;
-		for (int w = 0; w < workloads.size(); w++)
-			m = std::max(workloads[w]->getCheckTimeout(), m);
-		return m;
+void CompoundWorkload::addFailureInjection(WorkloadRequest& work) {
+	if (!work.runFailureWorkloads || !FLOW_KNOBS->ENABLE_SIMULATION_IMPROVEMENTS) {
+		return;
+	}
+	// Some common workloads won't work with failure injection workloads
+	for (auto const& w : workloads) {
+		auto desc = w->description();
+		if (desc == "ChangeConfig") {
+			return;
+		} else if (desc == "SaveAndKill") {
+			return;
+		}
+	}
+	auto& factories = IFailureInjectorFactory::factories();
+	DeterministicRandom random(sharedRandomNumber);
+	for (auto& factory : factories) {
+		auto workload = factory->create(*this);
+		while (workload->add(random, work, *this)) {
```

Comment:
I'm guessing that here we may need to create a maximum number of times that a particular failure injection workload can be added (even though it will be extremely rare to add >3 of the same workload, assuming the workload doesn't override `FailureInjectionWorkload::add`).